src/TableFunctions/TableFunctionS3Distributed.cpp

```diff
@@ -0,0 +1,143 @@
+#include <thread>
+#include <Common/config.h>
+#include "DataStreams/RemoteBlockInputStream.h"
+#include "Interpreters/ClientInfo.h"
+#include "Parsers/ASTExpressionList.h"
+#include "Parsers/ASTFunction.h"
+#include "Parsers/IAST_fwd.h"
+#include "Processors/Sources/SourceFromInputStream.h"
+#include "Storages/StorageS3Distributed.h"
+
+#if USE_AWS_S3
+
+#include <IO/S3Common.h>
+#include <Storages/StorageS3.h>
+#include <Interpreters/evaluateConstantExpression.h>
+#include <Interpreters/Context.h>
+#include <TableFunctions/TableFunctionFactory.h>
+#include <TableFunctions/TableFunctionS3.h>
+#include <TableFunctions/TableFunctionS3Distributed.h>
+#include <TableFunctions/parseColumnsListForTableFunction.h>
+#include <Parsers/ASTLiteral.h>
+#include "registerTableFunctions.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+void TableFunctionS3Distributed::parseArguments(const ASTPtr & ast_function, const Context & context)
+{
+    /// Parse args
+    ASTs & args_func = ast_function->children;
+
+    if (args_func.size() != 1)
+        throw Exception("Table function '" + getName() + "' must have arguments.", ErrorCodes::LOGICAL_ERROR);
+
+    ASTs & args = args_func.at(0)->children;
+
+    if (args.size() < 4 || args.size() > 7)
+        throw Exception("Table function '" + getName() + "' requires 4 to 7 arguments: cluster, url," + 
+            "[access_key_id, secret_access_key,] format, structure and [compression_method].",
```

Comment:
This signature seem to be not very solid.
If size = 4, it will be `cluster, url, format, structure`
If size = 5, it will be `cluster, url, format, structure, compression_method`
If size = 6, it will be `cluster, url, access_key_id, secret_access_key, format, structure`
If size = 7, it will be `cluster, url, access_key_id, secret_access_key, format, structure, compression_method`
Now it works, but we won't be ever able to add another one parameter.

Reply:
I don't really know how to rewrite this, because all except cluster were inherited from `s3` table function